PVC creation test

pohly opened the pull request:
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

I started using the experimental pod-startup test to measure how quickly provisioning of volumes via PVC can be done. Some tweaks to the test were necessary - see individual commit messages for details.

**Special notes for your reviewer**:
```release-note
NONE
```

pohly commented on clusterloader2/testing/experimental/storage/pod-startup/config.yaml:
```diff
@@ -7,12 +7,22 @@
-  {{$VOLUMES_PER_POD := .VOLUMES_PER_POD}}
+  {{$VOLUMES_PER_POD := DefaultParam .VOLUMES_PER_POD 1}}
   {{$VOLUME_TEMPLATE_PATH := .VOLUME_TEMPLATE_PATH}}
+  # Set this to false if apiserver, kube-controller-manager and kube-scheduler
```
But you do for the other components?

I need to double-check whether I can drop the "insecure-port: '8080'" in my local cluster config. It might be that I added it pro-actively.

pohly commented on clusterloader2/testing/experimental/storage/pod-startup/config.yaml:
```diff
@@ -7,12 +7,22 @@
+  {{$VOLUMES_PER_POD := DefaultParam .VOLUMES_PER_POD 1}}
   {{$VOLUME_TEMPLATE_PATH := .VOLUME_TEMPLATE_PATH}}
+  # Set this to false if apiserver, kube-controller-manager and kube-scheduler
+  # do not export metrics data via the insecure http ports.
```
I ran against a cluster where "kubeadm" locked down access to metrics via http and then gathering those metrics in clusterload2 failed because it couldn't connect to the ports.

The code which pulls metrics uses the insecure ports: https://github.com/kubernetes/kubernetes/blob/e799c852fb337be414fa2dae982fed69a67806fe/test/e2e/framework/metrics/metrics_grabber.go#L35-L44

pohly replied:
kube-apiserver indeed works without the insecure port.

But a missing insecure port for kube-controller-manager causes the test to fail:
```
...
I1013 17:00:22.539742  220515 resource_gather_worker.go:69] Get container kube-controller-manager-pmem-csi-pmem-govm-master/kube-controller-manager usage on node pmem-csi-pmem-govm-master. CPUUsageInCores: 0.011748654, MemoryUsageInBytes: 0, MemoryWorkingSetInBytes: 44888064
I1013 17:00:22.539758  220515 resource_gather_worker.go:69] Get container kube-proxy-7474b/kube-proxy usage on node pmem-csi-pmem-govm-master. CPUUsageInCores: 0.000631058, MemoryUsageInBytes: 0, MemoryWorkingSetInBytes: 13516800
E1013 17:00:30.651327  220515 metrics_for_e2e.go:79] MetricsForE2E: metricsGrabber failed to grab some of the metrics: Errors while grabbing metrics: [error waiting for controller manager pod to expose metrics: timed out waiting for the condition; an error on the server ("unknown") has prevented the request from succeeding (get pods kube-controller-manager-pmem-csi-pmem-govm-master:10252)]
...
```

wojtek-t replied:
OK - looked into the code and I know what is happening (I was wondering why this works for GKE where we don't even expose access to pods)

And apparently - we're grabbing those metrics via listing pods and if we don't find pods, we simply ignore them:
https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/metrics/metrics_grabber.go#L98

This code will have to be rewritten anyway (it e.g. doesn't support HA setups).

But in the meantime, the correct way to solve this would be to:
- introducing new features similar to SupportGrabMetricsFromKubelets in provider package:
https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/provider/provider.go#L52
- using this in the measurement similarly as we do for kubelets:
https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/measurement/common/metrics_for_e2e.go#L63

That should address your problem and allow you to run all the other measurements that are actually useful.

pohly commented on clusterloader2/testing/experimental/storage/pod-startup/config.yaml:
```diff
@@ -57,9 +69,11 @@ steps:
 {{ if $PROVISION_VOLUME }}
 # Provision volumes
-- phases:
+- name: Provisioning volumes
```
Sure.

pohly commented on clusterloader2/testing/experimental/storage/pod-startup/volume-types/persistentvolume/pvc.yaml:
```diff
@@ -5,11 +5,13 @@ metadata:
     app: {{.Name}}
     group: {{.Group}}
+  annotations:
+    volume.beta.kubernetes.io/storage-provisioner: {{.Provisioner}}
```
Gotcha :sob: I assumed that the actual value would still get added, but I didn't test that. Now I tested and the volume provisioning gets stuck because it remains empty. I'll revise this.

pohly commented on clusterloader2/testing/experimental/storage/pod-startup/config.yaml:
```diff
@@ -7,12 +7,13 @@
 
 # Test Variales
-  {{$PODS_PER_NODE := .PODS_PER_NODE}}
+  {{$PODS_PER_NODE := DefaultParam .PODS_PER_NODE 1}}
```
I actually wanted to ask about this: `DefaultParam` is supposed to make it possible to provide those values via environment variables instead of overrides, right?

I'm following the pattern from `NODES_PER_NAMESPACE` here, but when I tested it, I found that the env variables are ignored. Is that perhaps because they don't start with `CL2_`?

That's broken then also for the other variables, something that I should fix before repeating the same mistake.

wojtek-t replied:
Yeah - only the ones prefixed with CL2_ are propagated down the stack (I can't remember the exact motivation for that, but there was a good reason to not propagate everything).

pohly replied:
I decided to stay with the current naming, i.e. env variables continue to not work...

pohly commented:
@wojtek-t : can you have another look? I rebased and cleaned up the commits.

I kept the "disable metrics gather" option. It's not needed anymore to avoid the test failure, but I found it useful nonetheless when I just wanted to see whether provisioning works.

wojtek-t commented on clusterloader2/testing/experimental/storage/pod-startup/config.yaml:
```diff
@@ -7,12 +7,20 @@
+  {{$VOLUMES_PER_POD := DefaultParam .VOLUMES_PER_POD 1}}
   {{$VOLUME_TEMPLATE_PATH := .VOLUME_TEMPLATE_PATH}}
+  # Set this to false if metrics data is not needed.
+  {{$GATHER_METRICS := DefaultParam .GATHER_METRICS true}}
```
I still believe this should be removed due to reasons discussed in the other thread (in particular api-call latencies SLO is measured behind this param and it should be measured).

pohly replied:
This depends on why you run the test. If it is to measure performance, then yes. But I was also using it to trigger a race condition (https://github.com/kubernetes-csi/external-provisioner/issues/507) and then metrics gathering is just unnecessary overhead. In particular in a large cluster it can take time as it walks through various pods.

wojtek-t replied:
I think there shouldn't be an option for that, but ok - let's not block this PR on it.

wojtek-t commented on clusterloader2/testing/experimental/storage/pod-startup/config.yaml:
```diff
@@ -7,12 +7,20 @@
+  # the PV controller must update each PVC before external-provisioner
+  # starts to work on it, which limits the volume creation rate.
+  # With it, volume creation starts immediately.
+  {{$PROVISIONER := DefaultParam .PROVISIONER ""}}
```
What happens when this is set to empty string?

pohly replied:
Due to the if check, the annotation is not set and the volume controller must add it after the PVC got created.

wojtek-t commented on clusterloader2/testing/experimental/storage/pod-startup/volume-types/persistentvolume/pvc.yaml:
```diff
@@ -5,11 +5,15 @@ metadata:
   labels:
     app: {{.Name}}
     group: {{.Group}}
+{{ if .Provisioner }}
```
Hmm - how does "ifs" for string work?

pohly replied:
Apparently it checks for non-empty, i.e. empty string = false.

wojtek-t replied:
Interesting - but that's fine :)

pohly commented:
/hold

Skipping the PV controller for provisioning but then still waiting for it to bind volumes is inconsistent.

I'll change it so that it's either used for setting the annotation and binding or not relied upon at all.

pohly commented:
/hold cancel

Now volume creation can be tested with and without dependency on the PV controller.

pohly commented:
/test pull-perf-tests-1000-adhoc

pohly commented:
/test pull-perf-tests-1000-adhoc

pohly commented:
/retest

wojtek-t commented on clusterloader2/testing/experimental/storage/pod-startup/config.yaml:
```diff
@@ -71,20 +95,32 @@ steps:
+      Params:
+        desiredPVCount: {{$totalVols}}
+        apiVersion: v1
+        provisioner: {{$EXPECTED_PROVISIONER}}
```
I think we want to pass labelSelector too (as in other waits).

pohly replied:
Hmm, I wonder whether `WaitForPVOptions` should even have a `Selector` field. That was added in 21d99e9d0 but I don't see what it is good for.

wojtek-t replied:
Can we remove it then?

pohly replied:
I'm not sure - perhaps it has some purpose with other provisioners that I just don't know about. Let's keep it.

wojtek-t replied:
It seems suspicious to have it set in some "waits" and not in others.

But OK - these tests are marked as experimental anyway for now ...

wojtek-t reviewed (commented):
Couple small-ish comments.

pohly reviewed (commented):
@wojtek-t I've pushed an update (all changes in the last commit), please take another look.

wojtek-t reviewed (commented):
Please also squash commits.

pohly commented:
> Please also squash commits.

Each commit is self-contained, but I can also squash - done.

wojtek-t commented:
/lgtm
/approve
